source/_posts/2015-03-26-creating-a-dynamic-single-page-app-for-our-genome-team-using-react.markdown

```diff
@@ -0,0 +1,96 @@
+
+![screenshot](http://cl.ly/image/310O3U3J3T2c)
+
+The backend was organized, modular, and interfaced seamlessly with the Artsy API. However, the single page app needs of our customer were met with a multitude of javascript-heavy files declaring global event listeners, needless to say, these were clunky and difficult to maintain. When we were tasked with resurrecting this app by fixing the critical bugs preventing it from being adopted, we found that simple requests took much more time than they should, and often resulted in a shamefully hacky fix. We intended to contribute as little work as possible up front in order to get the app out the door as soon as possible, and then to refactor retroactively. After a month of frustration with no end to bugs in sight, we decided to stop trying to patch our leaky roof and build a new one.
+
+### Choosing a suitable framework 
+
+We spent a day researching different front end frameworks to see what would best fit our needs. Our requirements were:
+- A robust view layer that could work on top of our already-solid Rails backend
+- Performant enough for an interaction-heavy single-page app, with hundreds of editable fields autosaving on change
+- We preferred light, streamlined frameworks, and valued freedom over unnecessary structure
+ 
+After deliberating with other members of our team and the vast online community, we were introduced to [React](http://facebook.github.io/react/), Facebook's view layer framework. We chose React because it provides much-needed structure and support for components in a single page app, without too much boilerplate. 
+ 
+Our plan was to eventually replace all of the existing `*.haml.erb` templates and global coffeescript mixins with discrete React components. We used the [react-rails](https://github.com/reactjs/react-rails) gem, which easily integrates React components with Rails views.
+ 
+In line with the React tutorial, we began by breaking up our UI into functional and visual components. For each component we found the relevant haml template, converted it into jsx and React components using dummy data, and eventually updated it to accept the correct props/state from our top-level component which did all of the dynamic fetching and saving. Then we deleted the associated haml and javascript code and breathed a sigh of relief.
+ 
+### Thinking the React way
```

Comment:
Seems like capitalization is not consistent. Maybe these should all be React Way or Drag and Drop